milli/src/search/new/query_graph.rs

```diff
@@ -0,0 +1,458 @@
+
+        // TODO: we could consider generalizing to 4,5,6,7,etc. ngrams
+        let (mut prev2, mut prev1, mut prev0): (Vec<u16>, Vec<u16>, Vec<u16>) =
+            (vec![], vec![], vec![root_node]);
+
+        let original_terms_len = terms.len();
+        for term_idx in 0..original_terms_len {
+            let mut new_nodes = vec![];
+            let new_node_idx = add_node(
+                &mut nodes_data,
+                QueryNodeData::Term(LocatedQueryTermSubset {
+                    term_subset: QueryTermSubset::full(Interned::from_raw(term_idx as u16)),
+                    positions: terms[term_idx].positions.clone(),
+                    term_ids: term_idx as u8..=term_idx as u8,
+                }),
+            );
+            new_nodes.push(new_node_idx);
+
+            if !prev1.is_empty() {
+                if let Some(ngram) =
+                    query_term::make_ngram(ctx, &terms[term_idx - 1..=term_idx], &nbr_typos)?
+                {
+                    let ngram_idx = add_node(
+                        &mut nodes_data,
+                        QueryNodeData::Term(LocatedQueryTermSubset {
+                            term_subset: QueryTermSubset::full(ngram.value),
+                            positions: ngram.positions,
+                            term_ids: term_idx as u8 - 1..=term_idx as u8,
+                        }),
+                    );
+                    new_nodes.push(ngram_idx);
+                }
+            }
+            if !prev2.is_empty() {
+                if let Some(ngram) =
+                    query_term::make_ngram(ctx, &terms[term_idx - 2..=term_idx], &nbr_typos)?
+                {
+                    let ngram_idx = add_node(
+                        &mut nodes_data,
+                        QueryNodeData::Term(LocatedQueryTermSubset {
+                            term_subset: QueryTermSubset::full(ngram.value),
+                            positions: ngram.positions,
+                            term_ids: term_idx as u8 - 2..=term_idx as u8,
+                        }),
+                    );
+                    new_nodes.push(ngram_idx);
+                }
+            }
+            (prev0, prev1, prev2) = (new_nodes, prev0, prev1);
+        }
+
+        let root_node = Interned::from_raw(root_node);
+        let end_node = Interned::from_raw(end_node);
+        let mut nodes = FixedSizeInterner::new(
+            nodes_data.len() as u16,
+            QueryNode {
+                data: QueryNodeData::Deleted,
+                predecessors: SmallBitmap::new(nodes_data.len() as u16),
+                successors: SmallBitmap::new(nodes_data.len() as u16),
+            },
+        );
+        for (node_idx, node_data) in nodes_data.into_iter().enumerate() {
+            let node = nodes.get_mut(Interned::from_raw(node_idx as u16));
+            node.data = node_data;
+        }
+        let mut graph = QueryGraph { root_node, end_node, nodes };
+        graph.build_initial_edges();
```

Comment:
Explanation: we have created all the nodes but did not connect them together. This function `build_initial_edges` create an edge between nodes representing adjacent terms in the user query.